tests/operators/test_python.py

```diff
@@ -844,17 +848,30 @@ def f(exit_code):
             assert ti.state == expected_state
 
 
+venv_cache_path = tempfile.mkdtemp(prefix="venv_cache_path")
```

Comment:
PythonVirtualenv/ExternalPython tests account for vast majority of the tests in Operator - mainly because we have to start new Python interpreters for every such tests but also because those tests are pretty comprehensive.

I changed the approach for those tests to get the Virtualenv tests  a "bit" (~20%) faster overall by swtiching the Venv tests to use caching by default. I also have a few tests where caching is disabled so we are testing that case as well, but in most cases we are not really loosing anything. Each distinct venv (as determined by hash of requiresments and other parameters in the Caching PR in #33355 by @jens-scheffler-bosch will be created at least at the first test that uses it - but then that step will be effectively skipped for every other test that uses same virtualenv specification.

The nice thing about that approach is that thanks to that I found and fixed a very nasty edge case #35252 - so it turned out that "caching in testing by default" is a good idea, as "caching" case is potentially more complex and has more edge cases than non-caching one ("you know cache invalidation is the most difficult thing in programming"). This #35252 was actually the very manifestation of cache invalidation not being performed :)

Reply:
Good idea to enable venv caching in tests - then we test more functionality and find nice race conditions as well :-D - can we enable x-dist also here and run them all in parallel? :-D

Reply:
> Good idea to enable venv caching in tests - then we test more functionality and find nice race conditions as well :-D - can we enable x-dist also here and run them all in parallel? :-D

Unfotunately not. 

Those are "classic" DB tests - they create Dag Run in DB and run task using Airflow's internal code which reads and updates the DB (and cleans it before /after) - so by the natur of it, it falls into "DB tests" rather than "Non-DB" tests where at most what we can do (and this is what I do already for some time) we split the test into "groups" (TEST_TYPE) and run the TEST_TYPES in parallell - each TEST_TYPE with it's own Database (via docker compose) and all the tests within each TEST_TYPE are executed sequentially because of the shared resource - DB they use.

But generally speaking this whole PR is about splitting the tests of ours into DB and Non-DB - and run the Non-DB tests with `pytest-xdist` precisely. Simply the Venv and External tests we are talking about are the "DB" ones.

BTW. Part of this change is documentation explaining all that and I just submitted a change explaining it - with a few TODOs for examples and more explanation so I'd really appreciate good review of those as well (and of course open to get some feedback/improvement ideas on the whole thing):

The draft doc is in TESTING.rst in this PR - see https://github.com/apache/airflow/pull/35160/files#diff-35a395614be3ab2f31c88bff2e106b942c11e0416ad3a2fbbcd8223cd1387478R76 for the version I pushed just hour ago or so.

Reply:
And just tio update this answer 'No it's not possible TODAY". But I can easily imagine that some time in the future we might run another set of optimisations to improve performance sof some of our tests - and we remove DB requirement from them - either via mocking, or maybe even by adding "per test method" sqlite or in-memory database  (I've done both in the past). 

However - this is a different CI/DEV/Test optimisation step. This change which basically aims to not rewrite the tests and change their structure but merely split out those thests that do not need DB and add mechanism that will keep it that way.

And yeah. I am happy to work with others on the next steps too :). I believe that we should continuously optimize and refine our dev tooling - at least until we grow in terms of community and usage, because that's the only way we can keep it working :D. 

So yeah some day we might want to attempt to convert those Virtualenv/ExternalPython to non-DB tests :). One more thing though - I think we really need to make sure it's worth it.. There is a value in hitting real DB during those tests because they are testing "Real" integration with them. So we should have quite a good percentage of those. I think we have a little to much (9000 non-DB vs. 7000 DB sounds like should be 15000 non-DB  vs 1500 DB so having a clear split and numbers laid out our goal shoudl be likely to move some of the DB tets to become non-DB ones. 

I am definitely in that train :)

Reply:
Got your point - and of course in another PR and not tomorrow. I was on the "same train" like you (actually sitting in a train at moment of writing now :-D) - Yes and agree testing operators is DB related in most cases to IO needed task meta data. But as we have a system of concurrent execution, would be cool (mid/long term) to be able to run on more test on parallel cores. But of course isolation is key...

Reply:
Yeah. Isolation is a key.

And we cannot get isolation when some (most of the DB ones) of our tests do something like this:

```python
    @pytest.fixture(autouse=True)
    def base_tests_setup(self, request, create_task_instance_of_operator, dag_maker):
        self.dag_id = f"dag_{slugify(request.cls.__name__)}"
        self.task_id = f"task_{slugify(request.node.name, max_length=40)}"
        self.run_id = f"run_{slugify(request.node.name, max_length=40)}"
        self.ds_templated = self.default_date.date().isoformat()
        self.ti_maker = create_task_instance_of_operator
        self.dag_maker = dag_maker
        self.dag = self.dag_maker(self.dag_id, template_searchpath=TEMPLATE_SEARCHPATH).dag
        clear_db_runs()
        yield
        clear_db_runs()
```

This is actual code from `BasePythonTest` which is base for all the Python tests. if you see at the last three lines  ... all dag_runs will be cleared alwasys before and after every test. This is the implementation:

```python
def clear_db_runs():
    with create_session() as session:
        session.query(Job).delete()
        session.query(Trigger).delete()
        session.query(DagRun).delete()
        session.query(TaskInstance).delete()
```

This is of course key for the test to have proper setup/teardown but also it means that if it is run via `pytest-xdist` it wil randomly delete all the dag_runs while other tests are using the same database... (and likely most of the 7000 other DB tests is doing something very similar to achieve reproducibility).

Not good. Not good at all. 

And even if we could  fix it in some tests, there is no way we prevent someone sneaking in a fixture or setup_method with `session.query(DagRun).delete()` in order to "clean-up" before and after the test (which is good idea as long as it does not impact other tests running in paralllel) . Not when we merge 30 PRs a day. And one faulty test can start randomly break any other tests, chasing such a "bad apple" would be a nightmare.

That's why I think there are three reasonable approaches:

1) Split your DB tests in serialized test groups where tests are executed sequentially in a group and each group has their own DB. This is what we have now. In our self-hosted runners in CI we have up to 8 (this is how many CPUS we have) complete docker compose instances running - each of them having a separate DB and each of them running sequentially a subset of tests (TEST_TYPE). In Public Runners we have up to 2 of those (because we only have 2 CPUS). This powers our tests for years now and saves countless hours of build time utilising parallelism of those instances -  I implemented it long time ago :) 

2) separate out tests not use DB (this is what this PR aims for) and run on them with xdist and to add mechanism to keep it working for the future - detecting running DB tests in `xdist` environment. And good thing is that we can "graduate" some tests from DB to non-DB easily with this PR (just make them not use DB and remove the marker)

3) Give each test, their own DB. That's probably far too costly as an overhead (especially that with DB tests we want to - likely test all supported DBs). I attempted once to expand the approach from 1) and have a separate DB "per test module" rather than "per test group" - so instead of 12 or so TEST_TYPES we have now to have 200 or so separately started docker-compose instances with DB and CI containers (up to 8 at a time of course)- but the overhead for starting/stoping images and DBs was so huge that the tests took way more than now to complete (at least 5 times more I think)

So I believe 1) is a good middle-ground between parallelism and isolation and when we get far less number of DB tests by followng 2) route and moving more tests to be non-DB, we will get "optimal" solution. I think chasing "let's get the DB tests run with xdist` might not achieve the goal due to overhead/complexity/flakiness ,